### PR TITLE
Fix HBW NUMA identification in case of HMAT support

### PR DESCRIPTION
**Pull request by michalbiesek**
Based on #484 .
I recommend reviewing this commit by commit

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/488)

### CODE REVIEW

File: src/memkind_mem_attributes.c [hunk 1]
```diff
@@ -242,43 +240,56 @@ int get_mem_attributes_hbw_nodes_mask(struct bitmask **hbw_node_mask)
 
     err = hwloc_topology_init(&topology);
     if (MEMKIND_UNLIKELY(err)) {
-        numa_bitmask_free(*hbw_node_mask);
+        ;
```
Comment (bratpiorka) [inline src/memkind_mem_attributes.c:243]: empty statement?

File: src/memkind_mem_attributes.c [hunk 2]
```diff
@@ -291,18 +302,67 @@ int get_mem_attributes_hbw_nodes_mask(struct bitmask **hbw_node_mask)
...
+            log_err("No HBW Nodes for init node %d.", init_node->os_index);
+            goto free_node_arr;
+        }
+
+        //validate single NUMA Node condition
+        if (node_variant == NODE_VARIANT_SINGLE &&
+            vec_size > 1) {
+            log_err("Invalid Numa Configuration for Node %d", init_node->os_index);
+            status = MEMKIND_ERROR_RUNTIME;
+            for (i = 0; i < num_cpu; ++i) {
```
Comment (bratpiorka) [inline src/memkind_mem_attributes.c:333]: this is the same code as 320-323 - could you move it to "free_node_array" section?

File: src/memkind_mem_attributes.c [hunk 3]
```diff
@@ -291,18 +302,67 @@ int get_mem_attributes_hbw_nodes_mask(struct bitmask **hbw_node_mask)
...
+                if (VEC_CAPACITY(&node_arr[i]))
+                    VEC_DELETE(&node_arr[i]);
             }
+            goto free_node_arr;
         }
+
+        // populate memory attribute nodemask to all CPU's from initiator NUMA node
+        hwloc_bitmap_foreach_begin(i, init_node->cpuset)
+        int node = -1;
```
Comment (bratpiorka) [inline src/memkind_mem_attributes.c:342]: please add indentation here

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/488)* status: 0 of 11 files reviewed, 3 unresolved discussions (waiting on @bratpiorka and @PatKamin)

___
*[src/memkind_mem_attributes.c, line 243 at r2](https://reviewable.io/reviews/memkind/memkind/488#-MRt889KcT5i8KA3AwIH-r2-243:-MRtCWaRDApwxAiQ_yTT:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/d37b7446ba1ed1da794f01995b50c183764c349b/src/memkind_mem_attributes.c#L243)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

empty statement?
</blockquote></details>

Done.
___
*[src/memkind_mem_attributes.c, line 333 at r2](https://reviewable.io/reviews/memkind/memkind/488#-MRt889KcT5i8KA3AwIH-r2-333:-MRtCsBA1eT6bjmL8-oS:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/d37b7446ba1ed1da794f01995b50c183764c349b/src/memkind_mem_attributes.c#L333)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

this is the same code as 320-323 - could you move it to "free_node_array" section?
</blockquote></details>

Done.
___
*[src/memkind_mem_attributes.c, line 342 at r2](https://reviewable.io/reviews/memkind/memkind/488#-MRt889KcT5i8KA3AwIH-r2-342:-MRtDBwT9YM_TqWBDCCg:b766hk5) ([raw file](https://github.com/memkind/memkind/blob/d37b7446ba1ed1da794f01995b50c183764c349b/src/memkind_mem_attributes.c#L342)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

please add indentation here
</blockquote></details>

Astyle forbids it :/ after #430 it would be probably possible

**Review by PatKamin — COMMENTED** [review]
> Reviewed 7 of 7 files at r1, 4 of 4 files at r2, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/488)* status: all files reviewed, 6 unresolved discussions (waiting on @bratpiorka and @michalbiesek)

___
*[src/memkind_hbw.c, line 361 at r2](https://reviewable.io/reviews/memkind/memkind/488#-MRtH0Zq3sCh5J4QA778:-MRtH0Zq3sCh5J4QA779:b-eu27xz) ([raw file](https://github.com/memkind/memkind/blob/d37b7446ba1ed1da794f01995b50c183764c349b/src/memkind_hbw.c#L361)):*
> ```C
> }
> 
> static bool is_hmat_support(void)
> ```

supported
___
*[src/memkind_mem_attributes.c, line 107 at r2](https://reviewable.io/reviews/memkind/memkind/488#-MRtHulF1eILZKyrZcye:-MRtHulF1eILZKyrZcyf:b-m4ev1c) ([raw file](https://github.com/memkind/memkind/blob/d37b7446ba1ed1da794f01995b50c183764c349b/src/memkind_mem_attributes.c#L107)):*
> ```C
>                 for (i = 0; i < num_local_nodes; ++i) {
>                     if (local_nodes[i]->attr->numanode.local_memory == 0) {
>                         log_info("Node skipped - Node %d is without memory.", i);
> ```

nit: has no memory
___
*[src/memkind_mem_attributes.c, line 330 at r2](https://reviewable.io/reviews/memkind/memkind/488#-MRtMHiV54Iq5GSjl9Gy:-MRtMHiV54Iq5GSjl9Gz:bm4yl45) ([raw file](https://github.com/memkind/memkind/blob/d37b7446ba1ed1da794f01995b50c183764c349b/src/memkind_mem_attributes.c#L330)):*
> ```C
>         //validate single NUMA Node condition
>         if (node_variant == NODE_VARIANT_SINGLE &&
>             vec_size > 1) {
> ```

nit: can be in a single line

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 7 of 7 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/488)* status: all files reviewed, 3 unresolved discussions (waiting on @michalbiesek)

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/488)* status: 7 of 11 files reviewed, 1 unresolved discussion (waiting on @bratpiorka and @PatKamin)

___
*[src/memkind_hbw.c, line 361 at r2](https://reviewable.io/reviews/memkind/memkind/488#-MRtH0Zq3sCh5J4QA778:-MRtZ2qS856CVewt2vNc:bb6un1j) ([raw file](https://github.com/memkind/memkind/blob/d37b7446ba1ed1da794f01995b50c183764c349b/src/memkind_hbw.c#L361)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

supported
</blockquote></details>

Done
___
*[src/memkind_mem_attributes.c, line 107 at r2](https://reviewable.io/reviews/memkind/memkind/488#-MRtHulF1eILZKyrZcye:-MRtXFFO9KJJX-milLwN:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/d37b7446ba1ed1da794f01995b50c183764c349b/src/memkind_mem_attributes.c#L107)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

nit: has no memory
</blockquote></details>

Done.
___
*[src/memkind_mem_attributes.c, line 330 at r2](https://reviewable.io/reviews/memkind/memkind/488#-MRtMHiV54Iq5GSjl9Gy:-MRtZx3X00Wtp-9eyG5r:bb6un1j) ([raw file](https://github.com/memkind/memkind/blob/d37b7446ba1ed1da794f01995b50c183764c349b/src/memkind_mem_attributes.c#L330)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

nit: can be in a single line
</blockquote></details>

Done

**Review by PatKamin — COMMENTED** [review]
:lgtm:

> Reviewed 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/488#-:-MRt__9Z5lNSDRVZOwrf:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @michalbiesek)

**Review by bratpiorka — COMMENTED** [review]
:lgtm:

> Reviewed 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/488#-:-MRt_arK9CC8aDMFi7mv:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @michalbiesek)

**Comment by michalbiesek** [comment]
> nit: can be in a single line

Now done.

**Review by PatKamin — COMMENTED** [review]
> Reviewed 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/488)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @michalbiesek)

**Pull request merged**